### PR TITLE
Add improved parameter generator function for FIM

### LINKED ISSUE #369: FIM v2.0: Improve parameter generator function to be applied to every test
## Description

For most of our test, we generate their configuration parameters through a function called `generate_params` (fim.py). We need to improve its functionality to be able to use it for every test and cover every possible case.

At least, we should cover these:

- [x] To be able to assign a different value to a same attribute and wildcard in a list for different modes.
- [x] To be able to append new parameters to every existing configuration.
- [x] To be able to generate as many configurations as wanted.

### PR DESCRIPTION
**Pull request by vicferpoy**
Hi team, this closes #369 .

# Summary

In this PR we are improving `generate_params` function to be able to use it in every FIM test and make its use easier. 

### Function

`def generate_params(extra_params: dict = None, apply_to_all: Union[Sequence[Any], Generator[dict, None, None]] = None,
                    modes: list = None):`

In addition to this, every test has been adapted to this "new" function, removing unnecessary code and `fim.py` and `tools.py` have had their docstring changed to Numpy.

# Examples

```
    >>> generate_params(extra_params={'REPORT_CHANGES': {'report_changes': ['yes', 'no']}, 'MODULE_NAME': 'name'},
    ...                 modes=['realtime', 'whodata'])
    ([{'FIM_MODE': {'realtime': 'yes'}, 'REPORT_CHANGES': {'report_changes': 'yes'}, 'MODULE_NAME': 'name'},
      {'FIM_MODE': {'whodata': 'yes'}, 'REPORT_CHANGES': {'report_changes': 'no'}, 'MODULE_NAME': 'name'}],
     [{'fim_mode': 'realtime', 'report_changes': 'yes', 'module_name': 'name'},
      {'fim_mode': 'whodata', 'report_changes': 'no', 'module_name': 'name'}])


```
    >>> generate_params(extra_params={'MODULE_NAME': 'name'}, apply_to_all={'FREQUENCY': {'frequency': [1, 2]}},
    ...                 modes=['scheduled', 'realtime'])
    ([{'FIM_MODE': '', 'MODULE_NAME': 'name', 'FREQUENCY': {'frequency': 1}},
      {'FIM_MODE': {'realtime': 'yes'}, 'MODULE_NAME': 'name', 'FREQUENCY': {'frequency': 1}},
      {'FIM_MODE': '', 'MODULE_NAME': 'name', 'FREQUENCY': {'frequency': 2}},
      {'FIM_MODE': {'realtime': 'yes'}, 'MODULE_NAME': 'name', 'FREQUENCY': {'frequency': 2}}],
     [{'fim_mode': 'scheduled', 'module_name': 'name', 'frequency': {'frequency': 1}},
      {'fim_mode': 'realtime', 'module_name': 'name', 'frequency': {'frequency': 1}},
      {'fim_mode': 'scheduled', 'module_name': 'name', 'frequency': {'frequency': 2}},
      {'fim_mode': 'realtime', 'module_name': 'name', 'frequency': {'frequency': 2}}])
```
    >>> generate_params(extra_params={'LIST_OF_VALUES': {'list': [[1,2,3]]}, 'MODULE_NAME': 'name'},
    ...                 modes=['scheduled'])
    ([{'FIM_MODE': '', 'LIST_OF_VALUES': {'list': [1, 2, 3]}, 'MODULE_NAME': 'name'}],
     [{'fim_mode': 'scheduled', 'list_of_values': [1, 2, 3], 'module_name': 'name'}])

```

# Tests performed
These are the tests that had more critical changes. `frequency` tests failed due to `Timeout`, as expected since it is a known error.

```
========================================== test session starts ==========================================
platform linux -- Python 3.6.8, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 207 items                                                                                     

test_tags/test_tags.py .......................................................................... [ 35%]
..............................................                                                    [ 57%]
test_skip/test_skip.py .ssss.ssss.ssss..ssss.ssss.ssss.                                           [ 73%]
test_scan/test_scan_day.py ...                                                                    [ 74%]
test_scan/test_scan_day_and_time.py ..                                                            [ 75%]
test_scan/test_scan_time.py ...                                                                   [ 77%]
test_checks/test_hard_link.py ............                                                        [ 83%]
test_max_eps/test_max_eps.py ...............                                                      [ 90%]
test_report_changes/test_report_changes_and_diff.py ......                                        [ 93%]
test_report_changes/test_report_deleted_diff.py ......                                            [ 96%]
test_frequency/test_frequency.py FFFFFFFF                                                         [100%]

=============================================== FAILURES ================================================
```

### CODE REVIEW

File: test_wazuh/test_fim/test_report_changes/test_report_changes_and_diff.py [hunk 1]
```diff
@@ -30,10 +30,8 @@
 
 conf_params, conf_metadata = generate_params({'REPORT_CHANGES': {'report_changes': 'yes'},
```
Comment (crd1985) [inline test_wazuh/test_fim/test_report_changes/test_report_changes_and_diff.py]: Lacks extra_params key.

File: test_wazuh/test_fim/test_report_changes/test_report_deleted_diff.py [hunk 2]
```diff
@@ -39,10 +39,8 @@ def change_conf(report_value):
     """" Returns a new ossec configuration with a changed report_value"""
     conf_params, conf_metadata = generate_params({'REPORT_CHANGES': {'report_changes': report_value},
```
Comment (crd1985) [inline test_wazuh/test_fim/test_report_changes/test_report_deleted_diff.py]: Lacks extra_params key

File: test_wazuh/test_fim/test_restrict/test_restrict_valid.py [hunk 3]
```diff
@@ -33,8 +33,7 @@
 
 # configurations
 
-conf_params, conf_metadata = generate_params({'TEST_DIRECTORIES': directory_str},
-                                             {'test_directories': directory_str})
+conf_params, conf_metadata = generate_params({'TEST_DIRECTORIES': directory_str})
```
Comment (crd1985) [inline test_wazuh/test_fim/test_restrict/test_restrict_valid.py]:
```suggestion
conf_params, conf_metadata = generate_params(extra_params={'TEST_DIRECTORIES': directory_str})
```

**Comment by vicferpoy** [comment]
## Tests performed after refactor

```
========================================== test session starts ==========================================
platform linux -- Python 3.6.8, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 90 items                                                                                      

test_fim/test_basic_usage/test_basic_usage_baseline_generation.py ..                              [  2%]
test_fim/test_basic_usage/test_basic_usage_changes.py ......                                      [  8%]
test_fim/test_basic_usage/test_basic_usage_create_rt_wd.py ........................               [ 35%]
test_fim/test_basic_usage/test_basic_usage_create_scheduled.py ............                       [ 48%]
test_fim/test_basic_usage/test_basic_usage_delete_folder.py ......                                [ 55%]
test_fim/test_basic_usage/test_basic_usage_entries_match_path_count.py .                          [ 56%]
test_fim/test_basic_usage/test_basic_usage_move_dir.py ....FF.FF                                  [ 66%]
test_fim/test_basic_usage/test_basic_usage_move_file.py ...............                           [ 83%]
test_fim/test_basic_usage/test_basic_usage_rename.py ...F.F                                       [ 90%]
test_fim/test_basic_usage/test_basic_usage_starting_agent.py .........                            [100%]

=============================================== FAILURES ================================================

```

**Pull request merged**